File: clients/ios/Camelot/APIClient.swift
import Foundation
import CryptoKit

struct APIError: LocalizedError {
    let message: String
    var errorDescription: String? { message }
}

actor APIClient {
    private let baseURL: URL
    private let session: URLSession

    init() {
        let configured = Bundle.main.object(forInfoDictionaryKey: "APIBaseURL") as? String
        baseURL = URL(string: configured ?? "http://localhost:3100")!
        let configuration = URLSessionConfiguration.default
        configuration.httpCookieAcceptPolicy = .always
        configuration.httpShouldSetCookies = true
        configuration.timeoutIntervalForRequest = 10
        session = URLSession(configuration: configuration)
    }

    func health() async throws {
        let _: HealthResponse = try await request("/api/health")
    }

    func signUp(name: String, email: String, password: String) async throws {
        let body = ["name": name, "email": email, "password": password]
        let _: AuthResponse = try await request("/api/auth/sign-up/email", method: "POST", body: body)
    }

    func signIn(email: String, password: String) async throws {
        let body = ["email": email, "password": password]
        let _: AuthResponse = try await request("/api/auth/sign-in/email", method: "POST", body: body)
    }

    func listOrganizations() async throws -> [OrganizationResponse] {
        try await request("/api/auth/organization/list")
    }

    func createOrganization(name: String) async throws -> OrganizationResponse {
        let slug = name.lowercased()
            .replacingOccurrences(of: "[^a-z0-9]+", with: "-", options: .regularExpression)
            .trimmingCharacters(in: CharacterSet(charactersIn: "-")) + "-" + String(UUID().uuidString.prefix(6)).lowercased()
        return try await request("/api/auth/organization/create", method: "POST", body: ["name": name, "slug": slug])
    }

    func push(organizationID: String, deviceID: UUID, mutations: [SyncMutation]) async throws -> PushResponse {
        try await request("/api/sync/push", method: "POST", body: PushBody(organizationId: organizationID, deviceId: deviceID, mutations: mutations))
    }

    func pull(organizationID: String, cursor: String) async throws -> PullResponse {
        let encoded = organizationID.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? organizationID
        return try await request("/api/sync/pull?organizationId=\(encoded)&cursor=\(cursor)&limit=200")
    }

    func uploadVideo(
        organizationID: String,
        mediaID: UUID,
        fileURL: URL,
        revision: String? = nil,
        progress: @escaping @MainActor @Sendable (Int64) -> Void,
        shouldContinue: @escaping @MainActor @Sendable () -> Bool
    ) async throws -> Int64 {
        let values = try fileURL.resourceValues(forKeys: [.fileSizeKey])
        guard let fileSize = values.fileSize, fileSize > 0 else { throw APIError(message: "Recording file is empty or missing.") }
        let upload: UploadSession = try await request("/api/uploads", method: "POST", body: StartUploadBody(
            organizationId: organizationID,
            mediaId: mediaID,
            fileName: revision.map { "\(mediaID)-\($0).\(fileURL.pathExtension)" } ?? fileURL.lastPathComponent,
            contentType: fileURL.pathExtension.lowercased() == "mp4" ? "video/mp4" : "video/quicktime",
            totalBytes: fileSize
        ))
        if upload.status == "complete" { return Int64(fileSize) }

        let received = Set(upload.receivedParts)
        let handle = try FileHandle(forReadingFrom: fileURL)
        defer { try? handle.close() }
        var uploaded = Int64(0)
        for part in 1...upload.totalParts {
            guard await shouldContinue() else { throw APIError(message: "Upload paused while recording.") }
            let offset = UInt64((part - 1) * upload.chunkSize)
            let byteCount = min(upload.chunkSize, fileSize - Int(offset))
            if received.contains(part) {
                uploaded += Int64(byteCount)
                await progress(uploaded)
                continue
            }
            try handle.seek(toOffset: offset)
            guard let data = try handle.read(upToCount: byteCount), !data.isEmpty else { throw APIError(message: "Could not read recording part \(part).") }
            let digest = SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
            try await uploadPart(uploadID: upload.uploadId, part: part, data: data, checksum: digest)
            uploaded += Int64(data.count)
            await progress(uploaded)
        }
        let _: CompleteUploadResponse = try await request("/api/uploads/\(upload.uploadId.uuidString)/complete", method: "POST", body: EmptyBody())
        return uploaded
    }

    func createShare(organizationID: String, mediaID: UUID) async throws -> URL {
        let share: ShareResponse = try await request("/api/shares", method: "POST", body: ShareBody(organizationId: organizationID, mediaId: mediaID))
        guard let url = URL(string: share.url) else { throw APIError(message: "Server returned an invalid share URL.") }
        return url
    }

    private func uploadPart(uploadID: UUID, part: Int, data: Data, checksum: String) async throws {
        var value = URLRequest(url: baseURL.appending(path: "/api/uploads/\(uploadID.uuidString)/parts/\(part)"))
        value.httpMethod = "PUT"
        value.httpBody = data
        value.timeoutInterval = 120
        value.setValue("application/octet-stream", forHTTPHeaderField: "Content-Type")
        value.setValue(checksum, forHTTPHeaderField: "x-chunk-sha256")
        value.setValue(baseURL.absoluteString, forHTTPHeaderField: "Origin")
        for attempt in 1...3 {
            do {
                let (_, response) = try await session.data(for: value)
                guard let http = response as? HTTPURLResponse else { throw URLError(.badServerResponse) }
                if (200..<300).contains(http.statusCode) { return }
                if http.statusCode < 500 { throw APIError(message: "Could not upload recording part \(part) (HTTP \(http.statusCode)).") }
                throw URLError(.badServerResponse)
            } catch let error as APIError {
                throw error
            } catch {
                guard attempt < 3 else { break }
                try await Task.sleep(for: .seconds(attempt))
            }
        }
        throw APIError(message: "Upload paused. It will resume when the server is reachable.")
    }

    private func request<Response: Decodable, Body: Encodable>(
        _ path: String,
        method: String = "GET",
        body: Body? = Optional<String>.none
    ) async throws -> Response {
        var request = URLRequest(url: baseURL.appending(path: path))
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(baseURL.absoluteString, forHTTPHeaderField: "Origin")
        if let body { request.httpBody = try JSONEncoder().encode(body) }
        do {
            let (data, response) = try await session.data(for: request)
            guard let http = response as? HTTPURLResponse else { throw APIError(message: "Invalid server response") }
            guard (200..<300).contains(http.statusCode) else {
                let server = try? JSONDecoder().decode(ServerError.self, from: data)
                throw APIError(message: server?.message ?? "Server returned HTTP \(http.statusCode)")
            }
            return try JSONDecoder().decode(Response.self, from: data)
        } catch let error as APIError {
            throw error
        } catch {
            throw APIError(message: "Cannot reach \(baseURL.host ?? "server"). Check Wi-Fi and local-network permission.")
        }
    }
}

private struct HealthResponse: Decodable { let status: String }
private struct AuthResponse: Decodable { let token: String? }
private struct ServerError: Decodable { let message: String? }
struct OrganizationResponse: Decodable { let id: String; let name: String }
struct PushBody: Encodable { let organizationId: String; let deviceId: UUID; let mutations: [SyncMutation] }
struct SyncMutation: Encodable {
    let mutationId: UUID
    let entityId: UUID
    let entityType: String
    let operation: String
    let baseVersion: Int?
    let parentId: UUID?
    let payload: [String: String]
    let clientTimestamp: String

    init(mutationId: UUID, entityId: UUID, entityType: String, operation: String = "upsert", baseVersion: Int?, parentId: UUID?, payload: [String: String], clientTimestamp: String) {
        self.mutationId = mutationId; self.entityId = entityId; self.entityType = entityType; self.operation = operation
        self.baseVersion = baseVersion; self.parentId = parentId; self.payload = payload; self.clientTimestamp = clientTimestamp
    }
}
struct PushResponse: Decodable { let results: [MutationResult] }
struct MutationResult: Decodable {
    let mutationId: UUID
    let status: String
    let version: Int?
    let serverVersion: Int?
}
struct PullResponse: Decodable { let changes: [RemoteChange]; let cursor: String; let hasMore: Bool }
struct RemoteChange: Decodable {
    let entityType: String
    let entityId: UUID
    let version: Int
    let operation: String
    let parentId: UUID?
    let payload: [String: String]
}
private struct StartUploadBody: Encodable { let organizationId: String; let mediaId: UUID; let fileName: String; let contentType: String; let totalBytes: Int }
private struct UploadSession: Decodable { let uploadId: UUID; let chunkSize: Int; let totalParts: Int; let receivedParts: [Int]; let status: String }
private struct EmptyBody: Encodable {}
private struct CompleteUploadResponse: Decodable { let mediaId: UUID; let status: String }
private struct ShareBody: Encodable { let organizationId: String; let mediaId: UUID }
private struct ShareResponse: Decodable { let token: String; let url: String }

File: apps/web/src/app/api-client.ts
/* Port of APIClient.swift. Talks to the Hono API (apps/api): Better Auth email+password,
   the organization plugin, health, sync push/pull, chunked uploads and share links.
   Sessions are cookie based, so every request sends credentials. */

export class APIError extends Error {
  constructor(message: string, readonly status?: number) { super(message); this.name = "APIError"; }
}

export interface OrganizationResponse { id: string; name: string }
export interface SyncMutation {
  mutationId: string;
  entityId: string;
  entityType: string;
  operation: "upsert" | "delete";
  baseVersion: number | null;
  parentId: string | null;
  payload: Record<string, unknown>;
  clientTimestamp: string;
}
export interface MutationResult { mutationId: string; status: string; version?: number | null; serverVersion?: number | null }
export interface PushResponse { results: MutationResult[] }
export interface RemoteChange { entityType: string; entityId: string; version: number; operation: string; parentId: string | null; payload: Record<string, string> }
export interface PullResponse { changes: RemoteChange[]; cursor: string; hasMore: boolean }
export interface UploadSession { uploadId: string; chunkSize: number; totalParts: number; receivedParts: number[]; status: string }

const REQUEST_TIMEOUT_MS = 10_000;

/** Slug for a new organization, matching the iOS client ("FC Rovers" → "fc-rovers-a1b2c3"). */
export function organizationSlug(name: string, suffix = crypto.randomUUID().slice(0, 6).toLowerCase()): string {
  const base = name.toLowerCase().replace(/[^a-z0-9]+/g, "-").replace(/^-+|-+$/g, "");
  return `${base}-${suffix}`;
}

export function createAPIClient(baseURL: string) {
  const host = (() => { try { return new URL(baseURL || location.origin).host; } catch { return "server"; } })();

  async function request<T>(path: string, init: { method?: string; body?: unknown; timeoutMs?: number } = {}): Promise<T> {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), init.timeoutMs ?? REQUEST_TIMEOUT_MS);
    let response: Response;
    try {
      response = await fetch(`${baseURL}${path}`, {
        method: init.method ?? "GET",
        headers: { "Content-Type": "application/json" },
        body: init.body === undefined ? undefined : JSON.stringify(init.body),
        credentials: "include",
        signal: controller.signal,
      });
    } catch {
      throw new APIError(`Cannot reach ${host}. Check your connection and that the API is running.`);
    } finally { clearTimeout(timer); }
    const text = await response.text();
    const json = text ? safeJSON(text) : null;
    if (!response.ok) {
      const message = (json as { message?: string; error?: string } | null)?.message ?? (json as { error?: string } | null)?.error;
      throw new APIError(message ?? `Server returned HTTP ${response.status}`, response.status);
    }
    return json as T;
  }

  return {
    baseURL,
    health: () => request<{ status: string }>("/api/health"),
    signUp: (name: string, email: string, password: string) => request<unknown>("/api/auth/sign-up/email", { method: "POST", body: { name, email, password } }),
    signIn: (email: string, password: string) => request<unknown>("/api/auth/sign-in/email", { method: "POST", body: { email, password } }),
    signOut: () => request<unknown>("/api/auth/sign-out", { method: "POST", body: {} }),
    listOrganizations: () => request<OrganizationResponse[]>("/api/auth/organization/list"),
    createOrganization: (name: string) => request<OrganizationResponse>("/api/auth/organization/create", { method: "POST", body: { name, slug: organizationSlug(name) } }),
    push: (organizationId: string, deviceId: string, mutations: SyncMutation[]) => request<PushResponse>("/api/sync/push", { method: "POST", body: { organizationId, deviceId, mutations } }),
    pull: (organizationId: string, cursor: string) => request<PullResponse>(`/api/sync/pull?organizationId=${encodeURIComponent(organizationId)}&cursor=${cursor}&limit=200`),
    createShare: async (organizationId: string, mediaId: string) => {
      const share = await request<{ token: string; url: string }>("/api/shares", { method: "POST", body: { organizationId, mediaId } });
      if (!share.url) throw new APIError("Server returned an invalid share URL.");
      return share.url;
    },

    /** Chunked, resumable upload mirroring `uploadVideo` on iOS. Returns the number of bytes uploaded. */
    async uploadVideo(organizationId: string, mediaId: string, file: File, options: { revision?: string; onProgress?: (bytes: number) => void; shouldContinue?: () => boolean } = {}): Promise<number> {
      if (file.size <= 0) throw new APIError("Recording file is empty or missing.");
      const extension = file.name.split(".").pop() ?? "mp4";
      const upload = await request<UploadSession>("/api/uploads", {
        method: "POST",
        body: { organizationId, mediaId, fileName: options.revision ? `${mediaId}-${options.revision}.${extension}` : file.name, contentType: file.type || (extension === "mp4" ? "video/mp4" : "video/quicktime"), totalBytes: file.size },
      });
      if (upload.status === "complete") return file.size;
      const received = new Set(upload.receivedParts);
      let uploaded = 0;
      for (let part = 1; part <= upload.totalParts; part += 1) {
        if (options.shouldContinue && !options.shouldContinue()) throw new APIError("Upload paused while recording.");
        const offset = (part - 1) * upload.chunkSize;
        const chunk = file.slice(offset, Math.min(offset + upload.chunkSize, file.size));
        if (received.has(part)) { uploaded += chunk.size; options.onProgress?.(uploaded); continue; }
        const bytes = await chunk.arrayBuffer();
        const digest = Array.from(new Uint8Array(await crypto.subtle.digest("SHA-256", bytes)), (b) => b.toString(16).padStart(2, "0")).join("");
        await uploadPart(upload.uploadId, part, bytes, digest);
        uploaded += bytes.byteLength;
        options.onProgress?.(uploaded);
      }
      await request<unknown>(`/api/uploads/${upload.uploadId}/complete`, { method: "POST", body: {} });
      return uploaded;
    },
  };

  async function uploadPart(uploadId: string, part: number, bytes: ArrayBuffer, checksum: string) {
    for (let attempt = 1; attempt <= 3; attempt += 1) {
      try {
        const response = await fetch(`${baseURL}/api/uploads/${uploadId}/parts/${part}`, { method: "PUT", body: bytes, credentials: "include", headers: { "Content-Type": "application/octet-stream", "x-chunk-sha256": checksum } });
        if (response.ok) return;
        if (response.status < 500) throw new APIError(`Could not upload recording part ${part} (HTTP ${response.status}).`, response.status);
      } catch (error) {
        if (error instanceof APIError) throw error;
      }
      if (attempt < 3) await new Promise((resolve) => setTimeout(resolve, attempt * 1000));
    }
    throw new APIError("Upload paused. It will resume when the server is reachable.");
  }
}

export type APIClient = ReturnType<typeof createAPIClient>;

function safeJSON(text: string): unknown { try { return JSON.parse(text); } catch { return null; } }

export const errorMessage = (error: unknown) => (error instanceof Error ? error.message : String(error));
